File: typescript/day1.ts
import _ from "lodash";

export function parseInput(input: string): number[] {
  const lines = input.split(/\r?\n/);
  const depths = lines.map(x => parseInt(x));

  return depths;
}

function calculateIncreases(values: number[]): number {
  const pairs = _.zip(
    values.slice(1),
    values.slice(0, values.length - 1)
  );

  const increasesAndDecreases: boolean[] = pairs.map(
    ([current, prev]) => current! > prev!);

  const increases: boolean[] = increasesAndDecreases.filter(x => x);

  return increases.length;

}

export function part1(input: string): number {
  const depths = parseInput(input);

  return calculateIncreases(depths);
}

export function computePart2Sums(depths: number[]): number[] {
  const sums = _.zip(
    depths.slice(2, depths.length),
    depths.slice(1, depths.length - 1),
    depths.slice(0, depths.length - 2))
    .map(([a, b, c]) => a! + b! + c!)
    ;

    return sums;
}

export function part2(input: string): number {
  const depths = parseInput(input);

  const sums = computePart2Sums(depths);

  return calculateIncreases(sums);
}

File: python/day1.py
def part1(input):
    depths = [int(d) for d in input]

    return count_of_increases(depths)

def part2(input):
    depths = [int(d) for d in input]
    sums = [a + b + c for (a, b, c) in zip(depths[2:], depths[1:], depths)]

    return count_of_increases(sums)

def count_of_increases(measurements):
    pairs = zip(measurements[1:], measurements)
    depth_increases = [True for (cur, prev) in pairs if cur > prev]

    return len(depth_increases)


if __name__ == "__main__":
    with open('day1.txt') as f:
        input = f.readlines()

    print(f'Day 1, part 1: {part1(input)}')
    print(f'Day 1, part 2: {part2(input)}')